File: lib/validator.js
const { getCompanyDetails } = require('./freshdesk');

const TARGET_RELATIONSHIP_ID = process.env.TARGET_RELATIONSHIP_ID || '0-466-515';
const RELATIONSHIP_FIELD_KEY = process.env.FRESHDESK_RELATIONSHIP_FIELD_KEY || 'relationship_id';
const ESCALATION_TAG = process.env.ESCALATION_TAG || 'escalate-unresolved';

function companyMatchesRelationshipId(company) {
  if (!company) return false;

  const customFields = company.custom_fields || {};
  if (customFields[RELATIONSHIP_FIELD_KEY] === TARGET_RELATIONSHIP_ID) {
    return true;
  }

  // Fallback: some setups store the relationship id under a differently
  // named custom field. Scan all custom field values as a safety net.
  return Object.values(customFields).some((value) => value === TARGET_RELATIONSHIP_ID);
}

/**
 * Verify the ticket belongs to the target company (Relationship ID "0-466-515").
 * Fetches the company using `ticket.company_id`.
 */
async function isTargetCompany(ticket) {
  if (!ticket || !ticket.company_id) return false;

  const company = await getCompanyDetails(ticket.company_id);
  return companyMatchesRelationshipId(company);
}

/**
 * Normalize a Freshdesk tag value (array, comma-separated string, or null) to an array.
 */
function normalizeTags(tags) {
  if (!tags) return [];
  if (Array.isArray(tags)) return tags.map((t) => String(t).trim());
  return String(tags)
    .split(',')
    .map((t) => t.trim())
    .filter(Boolean);
}

/**
 * Check whether the escalation tag was newly added in this update.
 * If `previousTags` is not available, falls back to checking it's present
 * in the current tags (use this when the automation rule already fires
 * only on "tag added" for this specific tag).
 */
function hasEscalationTag(previousTags, currentTags) {
  const current = normalizeTags(currentTags);
  if (!current.includes(ESCALATION_TAG)) return false;

  if (previousTags === undefined || previousTags === null) {
    return true;
  }

  const previous = normalizeTags(previousTags);
  return !previous.includes(ESCALATION_TAG);
}

module.exports = {
  isTargetCompany,
  hasEscalationTag,
  TARGET_RELATIONSHIP_ID,
  ESCALATION_TAG,
};

File: lib/validator.py
import os

from lib.freshdesk import get_company_details

TARGET_RELATIONSHIP_ID = os.environ.get("TARGET_RELATIONSHIP_ID", "0-466-515")
RELATIONSHIP_FIELD_KEY = os.environ.get("FRESHDESK_RELATIONSHIP_FIELD_KEY", "cf_relationship_id")
ESCALATION_TAG = os.environ.get("ESCALATION_TAG", "escalate-unresolved")


def is_target_company(ticket):
    """Verify the ticket belongs to the target company (Relationship ID match)."""
    company_id = ticket.get("company_id")
    if not company_id:
        return False

    company = get_company_details(company_id)
    custom_fields = company.get("custom_fields") or {}

    if custom_fields.get(RELATIONSHIP_FIELD_KEY) == TARGET_RELATIONSHIP_ID:
        return True

    # Fallback: some accounts name the field differently; scan all values.
    return TARGET_RELATIONSHIP_ID in custom_fields.values()


def has_escalation_tag(previous_tags, current_tags):
    """Return True only if the escalation tag is newly added in this update."""
    previous_tags = previous_tags or []
    current_tags = current_tags or []

    return ESCALATION_TAG in current_tags and ESCALATION_TAG not in previous_tags
